awx/api/serializers.py

```diff
@@ -5481,41 +5526,43 @@ def check_peers_changed():
-            - instance peers matches peers in attrs
+            - instance peers does not match peers in attrs
             '''
             return self.instance and 'peers' in attrs and set(self.instance.peers.all()) != set(attrs['peers'])
 
         if not self.instance and not settings.IS_K8S:
             raise serializers.ValidationError(_("Can only create instances on Kubernetes or OpenShift."))
 
         node_type = get_field_from_model_or_attrs("node_type")
-        peers_from_control_nodes = get_field_from_model_or_attrs("peers_from_control_nodes")
-        listener_port = get_field_from_model_or_attrs("listener_port")
-        peers = attrs.get('peers', [])
-
-        if peers_from_control_nodes and node_type not in (Instance.Types.EXECUTION, Instance.Types.HOP):
-            raise serializers.ValidationError(_("peers_from_control_nodes can only be enabled for execution or hop nodes."))
 
         if node_type in [Instance.Types.CONTROL, Instance.Types.HYBRID]:
             if check_peers_changed():
                 raise serializers.ValidationError(
                     _("Setting peers manually for control nodes is not allowed. Enable peers_from_control_nodes on the hop and execution nodes instead.")
                 )
 
-        if not listener_port and peers_from_control_nodes:
-            raise serializers.ValidationError(_("Field listener_port must be a valid integer when peers_from_control_nodes is enabled."))
-
-        if not listener_port and self.instance and self.instance.peers_from.exists():
-            raise serializers.ValidationError(_("Field listener_port must be a valid integer when other nodes peer to it."))
-
-        for peer in peers:
-            if peer.listener_port is None:
-                raise serializers.ValidationError(_("Field listener_port must be set on peer ") + peer.hostname + ".")
-
         if not settings.IS_K8S:
             if check_peers_changed():
                 raise serializers.ValidationError(_("Cannot change peers."))
 
+        # cannot peer to self
+        peers_ids = [p.id for p in attrs.get('peers', [])]
+        if self.instance and self.instance.receptor_addresses.filter(id__in=peers_ids).exists():
+            raise serializers.ValidationError(_("Instance cannot peer to its own address."))
+
+        if self.instance and self.instance.receptor_addresses.all().exists():
+            instance_addresses = set(self.instance.receptor_addresses.all())
+            # cannot peer to an instance that is already peered to this instance
+            for p in attrs.get('peers', []):
+                if set(p.instance.peers.all()) & instance_addresses:
+                    raise serializers.ValidationError(_(f"Instance {p.instance.hostname} is already peered to this instance."))
+
+        # cannot peer to instance more than once
+        # compare length of set to original list to check for duplicates
+        peers_instances = [p.instance for p in attrs.get('peers', [])]
```

Comment:
You should probably do this block before the previous one, since it doesn't involve querying the db just checking the user's input.